tracer/src/Datadog.Trace/Debugger/Instrumentation/Registry/ProbeMetadataRegistry.cs

```diff
@@ -0,0 +1,108 @@
+// <copyright file="ProbeMetadataRegistry.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading.Tasks;
+using Datadog.Trace.Debugger.Expressions;
+using Datadog.Trace.Debugger.Helpers;
+using Datadog.Trace.Debugger.RateLimiting;
+
+namespace Datadog.Trace.Debugger.Instrumentation.Registry
+{
+    /// Acts as a registry of indexed <see cref="ProbeMetadataInfo"/>.
+    /// Each instrumented probe is given an index (hard-coded into the instrumented bytecode),
+    /// and this index is used to perform O(1) lookup for the corresponding ProbeMetadataInfo instance.
+    /// The reason the index is incremented and dictated by the native side is to support multi-AppDomain scenarios.
+    /// In these scenario, there will be multiple <see cref="EverGrowingRegistry{TPayload}.Items"/> arrays, one for each AppDomain.
+    /// In order for us to grab the same ProbeMetadataInfo across all of them, we need to dereference the same index,
+    /// because the same instrumented bytecode could execute in different AppDomains, and static fields are not shared across AppDomains.
+    internal class ProbeMetadataRegistry : EverGrowingRegistry<ProbeMetadataInfo>
+    {
+        private static readonly object InstanceLock = new();
+
+        private static ProbeMetadataRegistry _instance;
+        private static volatile bool _instanceInitialized;
+
+        public static ProbeMetadataRegistry Instance
```

Comment:
nitpick 1: Did you consider using `LazyInitializer.EnsureInitialized` or using a static constructor?
nitpick 2: I think you can remove the `volatile _instanceInitialized`, and use `Volatile.Read` and `VolatileWrite` on `_instance` for checking for null. 
You can resolve it if you want.

Reply:
Good point, I considered relying on `LazyInitializer` to make it more readable. I'll consider changing it.

Reply:
Done in 76aa6e38ec616c0be4fab8d8fd14c78331325de0.